### PR TITLE
Kube state metrics alerts

### PR DESCRIPTION
**Pull request by philbrookes**
@david-martin @aidenkeating

### CODE REVIEW

**Comment by david-martin** [comment]
@pat-cremin @brizrobbo  would appreciate your input on these initial set of alerts.

* *KubePodCrashLooping* - message format `Pod <namespace>/<pod> (<container>) is restarting <x> times / 5 minutes`
* *KubePodNotReady* - `Pod <namespace>/<pod> has been in a non-ready state for longer than an hour.`

These alerts will trigger for any pod in any namespace that has a specific label (monitoring-key:middleware) i.e. only the namespaces we're interested in monitoring.

I'd also be interested in your opinion on how quick should these alerts trigger. Right now, they'll only alert after 1 hour of the expression returning a non-zero value.

More info on the pod metrics these expressions are using can be seen here https://github.com/kubernetes/kube-state-metrics/blob/master/Documentation/pod-metrics.md

File: evals/roles/middleware_monitoring/files/kube_state_metrics_alerts.yml [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+spec:
+  groups:
+    - name: general.rules
+      rules:
+      - alert: KubePodCrashLooping
+        annotations:
+          message: Pod {{ $labels.exported_namespace }}/{{ $labels.exported_pod }} ({{ $labels.container }}) is restarting {{ printf "%.2f" $value }} times / 5 minutes.
+        expr: |
+          rate(kube_pod_container_status_restarts_total{job="kube-state-metrics"}[15m]) * on (namespace, exported_namespace) group_left(label_monitoring_key) kube_namespace_labels{label_monitoring_key="middleware"} * 60 * 5 > 0
+        for: 1h
```
Comment (david-martin) [inline evals/roles/middleware_monitoring/files/kube_state_metrics_alerts.yml]: Lets lower this to 15m

File: evals/roles/middleware_monitoring/files/kube_state_metrics_alerts.yml [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+        labels:
+          severity: critical
+      - alert: KubePodNotReady
+        annotations:
+          message: Pod {{ $labels.exported_namespace }}/{{ $labels.exported_pod }} has been in a non-ready state for longer than an hour.
+        expr: |
+          sum by(exported_pod, exported_namespace) (kube_pod_status_phase{phase=~"Pending|Unknown"} * on (namespace, exported_namespace) group_left(label_monitoring_key) kube_namespace_labels{label_monitoring_key="middleware"}) > 0
+        for: 1h
```
Comment (david-martin) [inline evals/roles/middleware_monitoring/files/kube_state_metrics_alerts.yml]: Lets lower this to 15m as well.
We can adjust these based on experience.
I think a faster alert time as a safe default is OK until we understand more about what triggers this alert

File: evals/roles/middleware_monitoring/tasks/main.yml [hunk 3]
```diff
@@ -11,10 +11,18 @@
 - name: Add monitoring label to namespace
   shell: oc patch ns {{ monitoring_namespace }} --patch '{"metadata":{"labels":{"{{ monitoring_label_name }}":"{{ monitoring_label_value }}"}}}'
 
+- name: "copy ksm template"
+  copy:
+    src: kube_state_metrics_alerts.yml
+    dest: /tmp/kube_state_metrics_alerts.yml
+
+- name: Create ksm alerts
+  shell: "oc create -f /tmp/kube_state_metrics_alerts.yml"
```
Comment (david-martin) [inline evals/roles/middleware_monitoring/tasks/main.yml]: might be better to use `oc apply` so it patches any differences

**Comment by pat-cremin** [comment]
@david-martin as per our conversation the severity of any alert will be apparent based on the experience of having the pod out of service. Assuming full HA any pod being in this state should not be a cause for concern. In this state though it will require attention eventually and perhaps immediate attention in a case where its fellow pods begin to follow suit or even if the state of the pod happens to mean the service is no longer HA capable. 

The disruption or potential disruption caused by a pod in a crashloopbackoff state will determine the eventual severity.  For now, I think you are right to treat all alerts in this state as critical until experience can tell you otherwise.

File: evals/roles/middleware_monitoring/tasks/main.yml (hunk 3 above)
Comment (david-martin) [reply evals/roles/middleware_monitoring/tasks/main.yml]: @philbrookes Can you add a `-n {{ monitoring_namespace }}` to this.
It caused below error when i manually deleted it and re-ran the installer

```fatal: [127.0.0.1]: FAILED! => {"changed": true, "cmd": "oc create -f /tmp/kube_state_metrics_alerts.yml", "delta": "0:00:00.181985", "end": "2019-02-07 15:35:56.761949", "msg": "non-zero return code", "rc": 1, "start": "2019-02-07 15:35:56.579964", "stderr": "Error from server (AlreadyExists): error when creating \"/tmp/kube_state_metrics_alerts.yml\": prometheusrules.monitoring.coreos.com \"ksm-alerts\" already exists", "stderr_lines": ["Error from server (AlreadyExists): error when creating \"/tmp/kube_state_metrics_alerts.yml\": prometheusrules.monitoring.coreos.com \"ksm-alerts\" already exists"], "stdout": "", "stdout_lines": []}
```

**Comment by david-martin** [comment]
@tremes @trepel would you mind giving this a spin and verifying the alerts?

File: evals/roles/kube_state_metrics/files/kube_state_metrics_alerts.yml [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+        annotations:
+          message: Pod {{ $labels.exported_namespace }}/{{ $labels.exported_pod }} ({{ $labels.container }}) is restarting {{ printf "%.2f" $value }} times / 5 minutes.
+        expr: |
+          rate(kube_pod_container_status_restarts_total{job="kube-state-metrics"}[15m]) * on (namespace, exported_namespace) group_left(label_monitoring_key) kube_namespace_labels{label_monitoring_key="middleware"} * 60 * 5 > 0
+        for: 15m
+        labels:
+          severity: critical
+      - alert: KubePodNotReady
+        annotations:
+          message: Pod {{ $labels.exported_namespace }}/{{ $labels.exported_pod }} has been in a non-ready state for longer than an hour.
```
Comment (david-martin) [inline evals/roles/kube_state_metrics/files/kube_state_metrics_alerts.yml]: @philbrookes the message here mentions an hour. Should change that to 15 minutes.

File: evals/roles/kube_state_metrics/files/kube_state_metrics_alerts.yml [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+  labels:
+    monitoring-key: middleware
+  name: ksm-alerts
+spec:
+  groups:
+    - name: general.rules
+      rules:
+      - alert: KubePodCrashLooping
+        annotations:
+          message: Pod {{ $labels.exported_namespace }}/{{ $labels.exported_pod }} ({{ $labels.container }}) is restarting {{ printf "%.2f" $value }} times / 5 minutes.
```
Comment (david-martin) [inline evals/roles/kube_state_metrics/files/kube_state_metrics_alerts.yml]: @philbrookes should we include 'over the last 15 minutes'

File: evals/roles/kube_state_metrics/files/kube_state_metrics_alerts.yml (hunk 4 above)
Comment (philbrookes) [reply evals/roles/kube_state_metrics/files/kube_state_metrics_alerts.yml]: nice spot, done.

File: evals/roles/kube_state_metrics/files/kube_state_metrics_alerts.yml (hunk 5 above)
Comment (philbrookes) [reply evals/roles/kube_state_metrics/files/kube_state_metrics_alerts.yml]: yeah, good idea, done.

**Comment by david-martin** [comment]
@philbrookes I'm good to merge once approved from either @tremes or @trepel.

**Comment by tremes** [comment]
Just for the record - the installation failed for me. See https://integreatly-qe-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/installation-pipeline/495/console. I am trying again

**Comment by philbrookes** [comment]
I have rebased this branch now.

**Comment by philbrookes** [comment]
This has just worked for me on my cluster, for verification:
- The rules object: [here](https://master.pbrookes-883b.openshiftworkshop.com/console/project/middleware-monitoring/edit/yaml?kind=PrometheusRule&name=ksm-alerts&group=monitoring.coreos.com&returnURL=project%2Fmiddleware-monitoring%2Fbrowse%2Fother%3Fkind%3DPrometheusRule%26group%3Dmonitoring.coreos.com)
- the alerts in prometheus [here](https://prometheus-route-middleware-monitoring.apps.pbrookes-883b.openshiftworkshop.com/alerts)

**Pull request merged**